src/main/java/net/minecraftforge/common/capabilities/CapabilityManager.java

```diff
@@ -61,31 +63,38 @@
         Objects.requireNonNull(storage,"Attempted to register a capability with no storage implementation");
         Objects.requireNonNull(factory,"Attempted to register a capability with no default implementation factory");
         String realName = type.getName().intern();
-        if (providers.containsKey(realName)) {
-            LOGGER.error(CAPABILITIES, "Cannot register capability implementation multiple times : {}", realName);
-            throw new IllegalArgumentException("Cannot register a capability implementation multiple times : "+ realName);
-        }
+        Capability<T> cap;
+
+        synchronized (providers)
+        {
+            if (providers.containsKey(realName)) {
+                LOGGER.error(CAPABILITIES, "Cannot register capability implementation multiple times : {}", realName);
+                throw new IllegalArgumentException("Cannot register a capability implementation multiple times : "+ realName);
+            }
 
-        Capability<T> cap = new Capability<>(realName, storage, factory);
-        providers.put(realName, cap);
+            cap = new Capability<>(realName, storage, factory);
+            providers.put(realName, cap);
+        }
 
         callbacks.getOrDefault(realName, Collections.emptyList()).forEach(func -> func.apply(cap));
     }
 
     // INTERNAL
-    private IdentityHashMap<String, Capability<?>> providers = new IdentityHashMap<>();
-    private IdentityHashMap<String, List<Function<Capability<?>, Object>>> callbacks = new IdentityHashMap<>();
+    private final IdentityHashMap<String, Capability<?>> providers = new IdentityHashMap<>();
+    private volatile IdentityHashMap<String, List<Function<Capability<?>, Object>>> callbacks;
```

Comment:
If I understand you correctly, volatile is not needed. ExecutorService.submit() already makes the changes visible.
`Actions in a thread prior to the submission of a Runnable to an Executor happen-before its execution begins. Similarly for Callables submitted to an ExecutorService. `

Reply:
whoops, I didn't see this. Since this volatile write only happens once, probably won't hurt to leave it.